docker/auth/auth.py

```diff
@@ -68,29 +60,29 @@ def resolve_repository_name(repo_name, insecure=False):
 
 
 def resolve_authconfig(authconfig, registry=None):
-    """Return the authentication data from the given auth configuration for a
-    specific registry. We'll do our best to infer the correct URL for the
-    registry, trying both http and https schemes. Returns an empty dictionnary
-    if no data exists."""
+    """
+    Returns the authentication data from the given auth configuration for a
+    specific registry. As with the Docker client, legacy entries in the config
+    with full URLs are stripped down to hostnames before checking for a match.
+    Returns None if no match was found.
+    """
     # Default to the public index server
-    registry = registry or INDEX_URL
-
-    # If it's not the index server there are three cases:
-    #
-    # 1. this is a full config url -> it should be used as is
-    # 2. it could be a full url, but with the wrong protocol
-    # 3. it can be the hostname optionally with a port
-    #
-    # as there is only one auth entry which is fully qualified we need to start
-    # parsing and matching
-    if '/v1/' not in registry:
-        registry = os.path.join(registry, 'v1/')
-    if not registry.startswith('http:') and not registry.startswith('https:'):
-        registry = 'https://' + registry
+    registry = convert_to_hostname(registry) if registry else INDEX_URL
 
     if registry in authconfig:
         return authconfig[registry]
-    return authconfig.get(swap_protocol(registry), None)
+
+    for key, config in six.iteritems(authconfig):
+        if convert_to_hostname(key) == registry:
+            return config
+
+    return None
+
+
+def convert_to_hostname(url):
+    url = url.replace('http://', '')
+    url = url.replace('https://', '')
+    return url.split('/', 1)[0]
```

Comment:
maybe use  https://docs.python.org/2/library/urlparse.html#urlparse.urlparse instead ? It should to be robust to unexpected parts of the url.

It's available as `six.moves.urllib.parse.urlparse`